### ISSUE TITLE
[macos] mounts hitting maxfiles limit on Catalina

### ISSUE THREAD

**Issue by ikalchev** [canonical/multipass#1933]
Hi! I recently started using multipass and it has been wonderful. Thanks!

**Describe the bug**
I am using multipass to spawn Ubuntu 20.10 as a build environment for a c++ project. The source code and its dependencies are mounted from the host using `multipass mount src ubuntu:/home/ubuntu/src`. The build would fail with seemingly random "file not found" errors for various header files. The files are actually there and the project builds if instead of mounting the files I copy them over in the VM.

**To Reproduce**
1. `multipass launch groovy --name ubuntu --cloud-init ./cloud-init.yaml -c 2 -m 8G`  // cloud init contains mostly package installs and env config.
2. `multipass mount src ubuntu:/home/ubuntu/src`. src contains a somewhat large amount of dependencies such as boost, openssl, rapidjson and others. The "file not found" errors are mostly for boost headers, but other files show up from time to time as well.
3. `multipass shell ubuntu`
4. `cd src && <run the build>`

Apologies that I am not able to provide the build command in its full, as it's a custom build system that ultimately configures CMake. However, I can get the "raw" clang++ command if necessary (that fails as well).

**Expected behavior**
The build should succeed similar to when the sources are copied instead of being mounted.

**Logs**
multipassd.log
```
[2021-01-21T10:36:31.683] [debug] [ubuntu] Trying SSH on 192.168.64.5:22
[2021-01-21T10:36:31.684] [debug] [ubuntu] process working dir ''
[2021-01-21T10:36:31.684] [info] [ubuntu] process program '/Library/Application Support/com.canonical.multipass/bin/hyperkit'
[2021-01-21T10:36:31.684] [info] [ubuntu] process arguments '-c, 2, -m, 4096M, -u, -A, -H, -U, 1d41c853-af58-33a7-ae54-990ce29417d8, -s, 0:0,hostbridge, -s, 2:0,virtio-net, -s, 5,virtio-rnd, -s, 31,lpc, -l, com1,autopty=/var/root/Library/Application Support/multipassd/vault/instances/ubuntu/pty,log=/Library/Logs/Multipass/ubuntu-hyperkit.log, -s, 1:0,virtio-blk,file:///var/root/Library/Application Support/multipassd/vault/instances/ubuntu/ubuntu-20.10-server-cloudimg-amd64.img?sync=os&buffered=1,format=qcow,qcow-config=discard=true;compact_after_unmaps=262144;keep_erased=262144;runtime_asserts=false, -s, 1:1,ahci-cd,/var/root/Library/Application Support/multipassd/vault/instances/ubuntu/cloud-init-config.iso, -f, kexec,/var/root/Library/Application Support/multipassd/vault/instances/ubuntu/ubuntu-20.10-server-cloudimg-amd64-vmlinuz-generic,/var/root/Library/Application Support/multipassd/vault/instances/ubuntu/ubuntu-20.10-server-cloudimg-amd64-initrd-generic,earlyprintk=serial console=ttyS0 root=/dev/vda1 rw panic=1 no_timer_check'
[2021-01-21T10:36:31.684] [info] [ubuntu] process state changed to Starting
[2021-01-21T10:36:31.686] [info] [ubuntu] process state changed to Running
[2021-01-21T10:36:31.695] [error] [ubuntu] Using fd 5 for I/O notifications
[2021-01-21T10:36:31.698] [info] [ubuntu] mirage_block_open: block_config = file:///var/root/Library/Application Support/multipassd/vault/instances/ubuntu/ubuntu-20.10-server-cloudimg-amd64.img?sync=os&buffered=1 and qcow_config = discard=true;compact_after_unmaps=262144;keep_erased=262144;runtime_asserts=false and stats_config = None

[2021-01-21T10:36:31.751] [info] [ubuntu] hyperkit: [INFO] Resized file to 30539 clusters (3908992 sectors)
[2021-01-21T10:36:31.751] [info] [ubuntu] hyperkit: [INFO] block recycler starting with keep_erased = 2048
[2021-01-21T10:36:31.752] [info] [ubuntu] hyperkit: [INFO] image has 0 free sectors and 30535 used sectors
[2021-01-21T10:36:31.752] [info] [ubuntu] mirage_block_open: block_config = file:///var/root/Library/Application Support/multipassd/vault/instances/ubuntu/ubuntu-20.10-server-cloudimg-amd64.img?sync=os&buffered=1 and qcow_config = discard=true;compact_after_unmaps=262144;keep_erased=262144;runtime_asserts=false and stats_config = None returning 0
mirage_block_stat

[2021-01-21T10:36:31.752] [info] [ubuntu] /var/root/Library/Application Support/multipassd/vault/instances/ubuntu/cloud-init-config.iso: fcntl(F_PUNCHHOLE) Operation not permitted: block device will not support TRIM/DISCARD
[2021-01-21T10:36:32.217] [info] [ubuntu] rdmsr to register 0x3a on vcpu 0
[2021-01-21T10:36:32.217] [info] [ubuntu] rdmsr to register 0x140 on vcpu 0
[2021-01-21T10:36:32.230] [info] [ubuntu] rdmsr to register 0x3a on vcpu 1
[2021-01-21T10:36:32.230] [info] [ubuntu] rdmsr to register 0x140 on vcpu 1
[2021-01-21T10:36:32.876] [info] [ubuntu] rdmsr to register 0x64e on vcpu 1
[2021-01-21T10:36:32.876] [info] [ubuntu] rdmsr to register 0x34 on vcpu 1
[2021-01-21T10:36:36.552] [info] [ubuntu] hyperkit: [INFO] Allocator: 30535 used; 0 junk; 0 erased; 0 available; 0 copies; 0 roots; 0 Copying; 0 Copied; 0 Flushed; 0 Referenced; max_cluster = 30538
[2021-01-21T10:36:36.552] [info] [ubuntu] hyperkit: [INFO] Allocator: file contains cluster 0 .. 30538 will enlarge file to 0 .. 31050
[2021-01-21T10:36:36.552] [info] [ubuntu] hyperkit: [INFO] resize: adding available clusters (Node ((x 30539) (y 31050) (l Empty) (r Empty) (h 1) (cardinal 512)))
[2021-01-21T10:36:36.753] [info] [ubuntu] hyperkit: [INFO] block recycler: triggering background flush: 30536 used; 0 junk; 0 erased; 511 available; 0 copies; 0 roots; 0 Copying; 0 Copied; 0 Flushed; 0 Referenced; max_cluster = 30539
[2021-01-21T10:36:42.264] [info] [sshfs-mounts] mounting /Users/kivan/Dev/repos/kivan_main_mbp => /home/ubuntu/kivan_main_mbp in ubuntu
[2021-01-21T10:36:42.264] [info] [sshfs-mounts] process program '/Library/Application Support/com.canonical.multipass/bin/sshfs_server'
[2021-01-21T10:36:42.264] [info] [sshfs-mounts] process arguments '192.168.64.5, 22, ubuntu, /Users/kivan/Dev/repos/kivan_main_mbp, /home/ubuntu/kivan_main_mbp, 502:-1,, 20:-1,'
[2021-01-21T10:36:42.264] [debug] [basic_process] starting: /Library/Application Support/com.canonical.multipass/bin/sshfs_server 192.168.64.5 22 ubuntu /Users/kivan/Dev/repos/kivan_main_mbp /home/ubuntu/kivan_main_mbp 502:-1, 20:-1,
[2021-01-21T10:36:54.628] [info] [ubuntu] hyperkit: [INFO] Allocator: 30792 used; 0 junk; 0 erased; 255 available; 0 copies; 0 roots; 0 Copying; 0 Copied; 0 Flushed; 0 Referenced; max_cluster = 30795
[2021-01-21T10:36:54.628] [info] [ubuntu] hyperkit: [INFO] Allocator: file contains cluster 0 .. 31050 will enlarge file to 0 .. 31562
[2021-01-21T10:36:54.628] [info] [ubuntu] hyperkit: [INFO] resize: adding available clusters (Node ((x 31051) (y 31562) (l Empty) (r Empty) (h 1) (cardinal 512)))
[2021-01-21T10:36:54.670] [info] [ubuntu] hyperkit: [INFO] Allocator: 31304 used; 0 junk; 0 erased; 255 available; 0 copies; 0 roots; 0 Copying; 0 Copied; 0 Flushed; 0 Referenced; max_cluster = 31307
[2021-01-21T10:36:54.670] [info] [ubuntu] hyperkit: [INFO] Allocator: file contains cluster 0 .. 31562 will enlarge file to 0 .. 32074
[2021-01-21T10:36:54.670] [info] [ubuntu] hyperkit: [INFO] resize: adding available clusters (Node ((x 31563) (y 32074) (l Empty) (r Empty) (h 1) (cardinal 512)))
[2021-01-21T10:36:54.711] [info] [ubuntu] hyperkit: [INFO] Allocator: 31816 used; 0 junk; 0 erased; 255 available; 0 copies; 0 roots; 0 Copying; 0 Copied; 0 Flushed; 0 Referenced; max_cluster = 31819
[2021-01-21T10:36:54.711] [info] [ubuntu] hyperkit: [INFO] Allocator: file contains cluster 0 .. 32074 will enlarge file to 0 .. 32586
[2021-01-21T10:36:54.711] [info] [ubuntu] hyperkit: [INFO] resize: adding available clusters (Node ((x 32075) (y 32586) (l Empty) (r Empty) (h 1) (cardinal 512)))
```
**Additional info**
- macOS 10.15.7
- `multipass version` - 1.5.0+mac
- `multipass info --all`
```
Name:           builder
State:          Running
IPv4:           192.168.64.7
Release:        Ubuntu 20.10
Image hash:     293e39b1e209 (Ubuntu 20.10)
Load:           1.00 1.00 1.00
Disk usage:     33.7G out of 96.7G
Memory usage:   166.6M out of 7.8G
```
- The build itself is CMake + ninja + clang 8.0

**Additional context**
Initially, I though it was something like this bug in WSL: https://stackoverflow.com/questions/50206890/intermittent-random-file-not-found-errors-under-windows-subsystem-for-linux. However, applying the workaround did not help (not using parallel builds).

When I run with `ninja -j1` to build serially, I get the error to be "stable", i.e. the "file not found" is always for the same file. Interesting enough, the "missing" file is different if I run the build from ninja and by executing the resulting command itself (i.e. `ninja -j1` vs `clang++ ...`).

Let me know if additional information is needed.

**Comment by Saviq** [comment]
Hi @ikalchev, thanks for the detailed bug report. We'll have to dig in and will likely need your help to debug. Stay tuned.

**Comment by ricab** [comment]
Hi @ikalchev,

Sorry you're experiencing this. We haven't seen this yet, so let me ask a bunch of questions with possible directions for investigation.

> When I run with ninja -j1 to build serially, I get the error to be "stable", i.e. the "file not found" is always for the same file.

OK, so there appears to be a deterministic problem accessing certain files. What happens when you try to read the contents of the files that the serial build fails on? And their parent directories? 

What about writing? Some builds need to generate or otherwise write files in the source directory and I wonder if that could be getting permission denied.

Are there any links in the original directory? Can you see them properly reflected in the mounted directory?

One thing you could try is to run `tree -pa` in the `src` directory before and after the build, or in copied and mounted dirs, and then diff the results.

Glad you're otherwise enjoying Multipass :slightly_smiling_face: Let us know of your findings, or if you find a way for us to reproduce.

**Comment by ikalchev** [comment]
Hey, thanks for the quick response, I appreciate it :)

> What happens when you try to read the contents of the files that the serial build fails on? And their parent directories?

Reading the "missing" file immediately after the fail works without a problem. In fact, include files from the same folder are found, i.e. we have:

```
#include <foo/bar.h>   // OK
#include <foo/zoo.h>  // Fails here
```

> What about writing?

Writing works, the previous steps of the build successfully write into the mounted tree.

> Are there any links in the original directory?

The only links are between shared libraries for versioning purposes. `ls -ltra` shows the links correctly, but I can't get to the link stage, so I cannot say if there is a problem there or not.

I did some more debugging using `strace -f -t -e trace=file clang++ ...` and only running one compile command (i.e. I ditched ninja). Snippet (had to change filenames, apologies):
```
...
17:21:02 openat(AT_FDCWD, "../foo/bar.h", O_RDONLY|O_CLOEXEC) = 7
17:21:02 readlink("/proc/self/fd/7", "/home/ubuntu/kivan_main_mbp/"..., 4096) = 174
17:21:02 openat(AT_FDCWD, "../foo/zoo.h", O_RDONLY|O_CLOEXEC) = -1 EPERM (Operation not permitted)
...
```

The compilation fails with a single error - "foo/zoo.h file not found". As you see, that fails with EPERM, but a file from the same folder was opened without a problem just before that. What is also interesting is that I can see EPERM for other files in the same folder AFTER the first EPERM for zoo.h happens.

It is also worth mentioning that I have McAfee Endpoint Security running on the host machine. I didn't have the time to try and disable (not sure if I can) that to see if it would fix the problem. I see that the compiler itself is trying to find every include file by appending it to every include directory, so it hits a lot of "expected" `file not found` errors while it tries to find its way around. Maybe this is triggering mcafee, but then that doesn't explain why compilations on the host machine don't fail.

**Comment by ricab** [comment]
Hi @ikalchev, ok, I am afraid I don't have much more to offer at this point, except to acknowledge that it does look like a bug on our side (perhaps some sort of race). Keeping this open to dig into later. Please let us know if you glean any more info.

**Comment by ikalchev** [comment]
Hi @ricab I have some new findings that I believe uncovered the issue.

TL;DR - macOS Catalina has a laughable `maxfiles` limit of 256, which is being hit inside the VM during a large compile. 

First, I managed to reproduce this without using any build system, so hopefully someone on macOS Catalina can try it. This simple python script simply opens (and never closes) a lot of files from the mounted partition:
```
import os

fs = [] # Holds opened files
dirstack = [
    # paths to mounted folders full of files. I used the include directories of the build dependencies, but any files will work.
]

# Traverse all folders and open every file until you hit an error.
for dirname, dirs, files in os.walk(dirstack.pop()):
    dirstack += [os.path.join(dirname, d) for d in dirs]
    for f in files:
        try:
            fs.append(open(os.path.join(dirname, f), "r"))
        except:
            print(len(fs))    # Print how many files we managed to open so far
            raise
```

Running this in the VM failed with:

```
$ python3 debug.py 
247
Traceback (most recent call last):
  File "debug.py", line 35, in <module>
PermissionError: [Errno 1] Operation not permitted: <filename...>
```

Some debugging later, I am pretty certain the issue is with the `maxfiles` limit in macOS Catalina being a mere 256:
```
~ >>> launchctl limit maxfiles
	maxfiles    256            unlimited 
```
The same limit inside the VM is much larger and setting it any higher doesn't help (which makes sense as the files are opened on the host machine as well).

I started looking for ways to increase that limit. Unfortunately, I briefly hit that https://github.com/canonical/multipass/issues/1398, which made me revert my changes. Fortunately, the issue resolved itself when I started the multipass "GUI". I will have another go at it over the weekend.

In the mean time, is it possible for Multipassd to use [`setrlimit`](https://pubs.opengroup.org/onlinepubs/009695399/functions/getrlimit.html) when spawning the VM? Can someone try to reproduce this so we are certain if this is in fact the issue?

**Comment by Saviq** [comment]
Hi @ikalchev, great digging! You should be able to add a limit in the instance in `rc.local` or such. The problem is, if two processes inside the instance would access the mount point at the same time, the limit would be hit again… So that's not a reliable fix.

**Comment by ikalchev** [comment]
I did manage to increase the limit to +65k. I then rebooted macOS and tried the script in the VM again.

This time, the script did not hit EPERM, but the process hanged. It hanged at the same number of files as before.

I restarted the VM and tried again - same. `ulimit` and launchctl both show the new limit. So the limit is definitely changing the behavior but why the hang?!

Any ideas on how to debug this further?

**Comment by ikalchev** [comment]
@Saviq can you elaborate on setting a limit in `rc.local`? I did increase the limit in macOS following [this](https://gist.github.com/devinrhode2/4cbf7f02a9701510d61f5be0515b8286)

**Comment by ricab** [comment]
Hi @ikalchev, very interesting finds. I managed to reproduce the issue with the python script you posted and we've been discussing. The script fails directly on the host as well, although with a more adequate `errno`. It's a pity that increasing the limit did not help (perhaps apple know their own limitations and have a reason to set it that low?). Do you get the same when running your script on the host after increasing the limit?

If your build is running multiple processes accessing multiple files in the mount, all of them will go to the same `sshfs_server`, which will hit the limit. Two things I can think of that you could try to overcome this:
1. throttle the build (number of parallel jobs)
2. split the source into multiple mounts, so that there are multiple `sshfs_server`s on the outside.

I believe the main issue for us then is that our `sshfs_server` should:
  - clearly log the occurrence 
  - ideally transfer the errno to inside the instance, in a way the OS saw the same thing and so gave the same `errno` to the process trying `open` call - to be investigated whether this is possible and how to achieve it, but setting a matching limit in the instance wouldn't be desirable, as we don't want to limit files outside mounts.

Some data points, for reference:
  - `errno`s relating to this
```
ENFILE 23 Too many open files in system
EMFILE 24 Too many open files
```
  - the python script produces the same error as on the mac after a `ulimit -Sn 128` inside the instance

**Comment by ikalchev** [comment]
@ricab when I increased the limit in macOS, the script did not fail on the host (it managed to open 10000 files before I stopped it). However, the same script just hanged inside the VM when opening the 247th file, so at least at first sight it doesn't seem like macOS is still using the old 256 limit. I will investigate if creating a new VM after the limit has been increased fixes this.

The compiler itself seems to keep open all of the 200+ headers for the duration of the compilation, apparently (per strace logs). So not much to do there.

Another venue of research is the sshfs server caching. Is it possible that the cache is getting messed up by this wonderful access pattern? With what options, if any, is the fuse mount done?

**Comment by ricab** [comment]
> Another venue of research is the sshfs server caching. Is it possible that the cache is getting messed up by this wonderful access pattern? With what options, if any, is the fuse mount done?

You mean the effects of the previous limit somehow lingering because of caching? I am not sure exactly what mechanism could be doing that, since you say you rebooted. Do you use any compilation cache inside the instance? A new instance would indeed be a good test. 

You can see the options that are passed to sshfs inside the instance. Not much departing from the default:

```
$ multipass exec primary -- ps aux | grep sshfs
root        2022  0.0  0.4  11008  4604 ?        Ss   11:42   0:00 sudo env LD_LIBRARY_PATH=/snap/multipass-sshfs/145/lib /snap/multipass-sshfs/145/bin/sshfs -o slave -o transform_symlinks -o allow_other -o Compression=no -o dcache_timeout=3 :/home/ricab /home/ubuntu/Home
root        2023  0.0  0.2 230172  2892 ?        Sl   11:42   0:00 /snap/multipass-sshfs/145/bin/sshfs -o slave -o transform_symlinks -o allow_other -o Compression=no -o dcache_timeout=3 :/home/ricab /home/ubuntu/Home
```

**Comment by ikalchev** [comment]
> Do you use any compilation cache inside the instance?

I meant the sshfs cache options as described here: http://manpages.ubuntu.com/manpages/groovy/en/man1/sshfs.1.html

**Comment by ricab** [comment]
> I meant the sshfs cache options as described here: http://manpages.ubuntu.com/manpages/groovy/en/man1/sshfs.1.html

OK, yeah the options we use are the ones in the command line above, so basically all defaults except a three-second directory-cache timeout.

I confirm the hang you mentioned after increasing the limit in macOS. This needs more time, but writing down what I found so far.

The first time I try the script, it hangs after opening the 256th file, while the daemon says 

```
[2021-02-01T12:19:16.971] [debug] [/Library/Application Support/com.canonical.multipass/bin/sshfs_server] [2021-02-01T12:19:16.971] [error] [sftp server] error occurred when replying to client: -1
```

Which comes out [here](https://github.com/canonical/multipass/blob/0e74ed7050fbafae933c1e84e6d633739a09bd46/src/sshfs_mount/sftp_server.cpp#L400). Inside the instance, the process gets into an "uninterruptible sleep" (state `D` in `ps`) waiting for disk. Which means it does not see any signals and can't be killed (no `^C`, no `pkill -SIGKILL`). At this point any subsequent attempt at looking into the mount gets into the same state. So if run again, the script will hang as soon as it tries to `os.walk` the directory.

The processes can be made to resume by stopping the mount (`multipass umount` in the host). They report an `OSError` exception in python which actually conveys an IO error. In the first process, it shows the 257th file it tried to open:

```
OSError: [Errno 5] Input/output error: '/home/ubuntu/Home/mydev/multipass/multipass-private/3rd-party/grpc/.github/lock.yml'
```

While the sshfs_server on the host usually crashes:

```
[2021-02-01T12:35:15.863] [debug] [sshfs-mounts] Stopping mount 'Home' in instance "primary"
[2021-02-01T12:35:15.863] [warning] [Qt] QProcess: Destroyed while process ("/Library/Application Support/com.canonical.multipass/bin/sshfs_server") is still running.
[2021-02-01T12:35:15.865] [error] [sshfs-mounts] There was an error with sshfs_server for instance "primary" with path 'Home': Crashed - program: /Library/Application Support/com.canonical.multipass/bin/sshfs_server; error: Process crashed
[2021-02-01T12:35:15.865] [warning] [sshfs-mounts] Mount 'Home' in instance "primary" has stopped unexpectedly: program: /Library/Application Support/com.canonical.multipass/bin/sshfs_server; error: Process crashed
```

All despite:

```
$ sysctl
Display all 1685 possibilities? (y or n)^C
$ sysctl kern.maxfilesperproc
kern.maxfilesperproc: 10240
$ sysctl kern.maxfiles
kern.maxfiles: 10240
$ launchctl limit maxfiles
        maxfiles    10240          10240
$ ulimit -n
10240
```

We need to handle such cases better, but I am still not sure how to actually get it to handle more files.

**Comment by ricab** [comment]
Another data point. *Manual* mounts do not suffer from this. The script hits a 1024 limit inside the instance first:
```
ubuntu@primary:~$ sshfs -o transform_symlinks -o Compression=no ricardo@192.168.64.1:./mydev/multipass/multipass-private /home/ubuntu/mppriv
Password:
ubuntu@primary:~$ python3 open-files.py /home/ubuntu/mppriv/3rd-party/
1021
Traceback (most recent call last):
  File "open-files.py", line 12, in <module>
OSError: [Errno 24] Too many open files: '/home/ubuntu/mppriv/3rd-party/grpc/tools/fuzzer/runners/ssl_server_fuzzer.sh'
Error in sys.excepthook:
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/apport_python_hook.py", line 55, in apport_excepthook
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 914, in _find_spec
  File "<frozen importlib._bootstrap_external>", line 1342, in find_spec
  File "<frozen importlib._bootstrap_external>", line 1314, in _get_spec
  File "<frozen importlib._bootstrap_external>", line 1443, in find_spec
  File "<frozen importlib._bootstrap_external>", line 1483, in _fill_cache
OSError: [Errno 24] Too many open files: '/home/ubuntu'

Original exception was:
Traceback (most recent call last):
  File "open-files.py", line 12, in <module>
OSError: [Errno 24] Too many open files: '/home/ubuntu/mppriv/3rd-party/grpc/tools/fuzzer/runners/ssl_server_fuzzer.sh'
ubuntu@primary:~$ ulimit -n
1024
```

**Comment by ricab** [comment]
I just retried on a Ubuntu host and got the [same thing](https://github.com/canonical/multipass/issues/1933#issuecomment-770837840), so this new 256 bound is a limitation in our mount approach somewhere, which will require some more investigation.

**Comment by ricab** [comment]
@ikalchev we're working on a fix for the hang you observe. In the meantime, you could workaround the issue by installing osxfuse (`brew install --cask osxfuse`) on the host and then running sshfs in the instance, like this:
```
mkdir <mount_dir>
sshfs <host_user>@<host_ip>:<host_directory> <mount_dir>
```

**Comment by townsend2010** [comment]
To recap, there are 3 cascading issues causing this.

1. macOS has a very low limit of the number (256) of files that can be opened simultaneously, but that can be increased by the user.
2. If the host OS limit is increased to more than 256 and more than 256 files are opened, then the mount "hangs" and cannot be salvaged until a `multipass unmount` is issued.
3. We use `libssh` for SFTP handling and they have an artificial hard-coded limit of 256 open file handles (#1950)

I'll address the issue in this bug by fixing `2.` above.

**Comment by ricab** [comment]
Hey @townsend2010, in my observation the mount hangs when hitting the limit in `libssh`. When it hits the OS limit, the opening process gets EPERM errno, while it should ideally get the original `ENFILE` or `EMFILE` from the `sshfs_server`. 

So yeah, 2 in here, but the hang itself would go in 3 (though improving error handling here may end up fixing it there as well?).

**Comment by townsend2010** [comment]
Hey @ricab,

> ...in my observation the mount hangs when hitting the limit in libssh.

Yep, that's `2.` in my list above.

> When it hits the OS limit, the opening process gets EPERM errno, while it should ideally get the original ENFILE or EMFILE...

Yes, that would be really great.  However, the SFTP protocol has very small set of errors that can be used with no way of setting more specific errno values.  As a matter of fact, `sshfs` is the one that reports these `EPERM`, etc. errors based on the limited SFTP error set.

**Comment by ricab** [comment]
Ah OK, I got confused by 2 being addressed here. I split into two issues to separate the handling of the two limits: (mac)OS and `libssh`.

**Comment by ikalchev** [comment]
> @ikalchev we're working on a fix for the hang you observe. In the meantime, you could workaround the issue by installing osxfuse (`brew install --cask osxfuse`) on the host and then running sshfs in the instance, like this:
> 
> ```
> mkdir <mount_dir>
> sshfs <host_user>@<host_ip>:<host_directory> <mount_dir>
> ```

Just tried it and it works great! Compile times are more than good enough for what I need. Thanks a lot for the workaround.

For others wanting to apply the workaround - make sure to enable remote login in Settings -> Sharing in macOS.

**Comment by Saviq** [comment]
Hi @ikalchev, could you please try this package?

https://github.com/canonical/multipass/pull/1990#issuecomment-789675165

It should fix the hang with more than 256 open files.
